File: app/src/main/java/net/cynreub/subly/domain/usecase/UpdateNextBillingDateUseCase.kt
package net.cynreub.subly.domain.usecase

import net.cynreub.subly.domain.model.BillingFrequency
import net.cynreub.subly.domain.model.Subscription
import net.cynreub.subly.domain.repository.SubscriptionRepository
import java.time.LocalDate
import javax.inject.Inject

class UpdateNextBillingDateUseCase @Inject constructor(
    private val subscriptionRepository: SubscriptionRepository
) {
    suspend operator fun invoke(subscription: Subscription) {
        val nextDate = calculateNextBillingDate(subscription)
        val updatedSubscription = subscription.copy(nextBillingDate = nextDate)
        subscriptionRepository.updateSubscription(updatedSubscription)
    }

    private fun calculateNextBillingDate(subscription: Subscription): LocalDate {
        val current = subscription.nextBillingDate
        return when (subscription.frequency) {
            BillingFrequency.WEEKLY -> current.plusWeeks(1)
            BillingFrequency.MONTHLY -> current.plusMonths(1)
            BillingFrequency.QUARTERLY -> current.plusMonths(3)
            BillingFrequency.SEMI_ANNUAL -> current.plusMonths(6)
            BillingFrequency.ANNUAL -> current.plusYears(1)
            BillingFrequency.CUSTOM -> current.plusMonths(1)
        }
    }
}

File: subly_ios/Subly/Subly/Domain/UseCases/UpdateNextBillingDateUseCase.swift
import Foundation

final class UpdateNextBillingDateUseCase {
    private let subscriptionRepository: SubscriptionRepository

    init(subscriptionRepository: SubscriptionRepository) {
        self.subscriptionRepository = subscriptionRepository
    }

    func execute(subscription: Subscription) async throws {
        let nextDate = calculateNextBillingDate(for: subscription)
        var updated = subscription
        updated.nextBillingDate = nextDate
        try await subscriptionRepository.updateSubscription(updated)
    }

    private func calculateNextBillingDate(for subscription: Subscription) -> Date {
        let cal = Calendar.current
        let current = subscription.nextBillingDate
        switch subscription.frequency {
        case .weekly:
            return cal.date(byAdding: .weekOfYear, value: 1, to: current)!
        case .monthly:
            return cal.date(byAdding: .month, value: 1, to: current)!
        case .quarterly:
            return cal.date(byAdding: .month, value: 3, to: current)!
        case .semiAnnual:
            return cal.date(byAdding: .month, value: 6, to: current)!
        case .annual:
            return cal.date(byAdding: .year, value: 1, to: current)!
        case .custom:
            return cal.date(byAdding: .month, value: 1, to: current)!
        }
    }
}
